source/client.c

```diff
@@ -612,6 +612,36 @@ static void s_attempt_reconnect(struct aws_task *task, void *userdata, enum aws_
 
         mqtt_connection_lock_synced_data(connection);
 
+        /**
+         * Check the state and determine if this reconnect is happening in the middle of a disconnect! If it is, then
+         * abort the reconnect and finish the disconnect (as the disconnect is now deadlocked thanks to the reconnect)
+         */
+        if (connection->synced_data.state == AWS_MQTT_CLIENT_STATE_DISCONNECTING) {
```

Comment:
I am not super confident in this code, though it does seem to work in all tests and fixes a deadlock I found in the `s_test_mqtt_connection_disconnect_while_reconnecting` test.

What seems to be the case in the test above is that the reconnect task happening in the middle of the disconnect causes a deadlock where the disconnect callback is never fired because the reconnect overrides it. If you do not complete the disconnect here, the test will hang either waiting for the disconnect to complete, or if you call ` MQTT_CLIENT_CALL_CALLBACK(connection, on_disconnect);` then it finishes the test but hangs waiting for memory to be cleaned up. Finishing the disconnect manually and ensuring everything is cleaned up seems to fix it in the tests I've run locally with CTEST.

Doing a straight cancel/abort of the reconnect task leads to this deadlock happening every time, while with this fix the deadlock never occurs.

However, I am not super confident that this is the best way to solve it...

Reply:
**OKAY** I have looked further into this and was right in the general idea but wrong in what was causing it, why it was occurring, and how the fix solved it.

The situation is pretty simple: We have a disconnection that happens right in the middle of a reconnect. When this happens, we do not want to reconnect anymore.

The problem is that the disconnect function, `aws_mqtt_client_connection_disconnect`, will ONLY set the state of the MQTT client to `AWS_MQTT_CLIENT_STATE_DISCONNECTING` if it is in the middle of a reconnect because there is no connection `slot` set, essentially making all it does change the state.
- Previously, this was not an issue because in the reconnect task, we would ALWAYS try to reconnect and this would trigger a connection, spin up a new channel and `slot`, see the state is set to disconnecting, and then shut everything down. However, this is what we want to avoid as that is a decent amount of work that we immediately throw away.

So, the solution: We have to check in the reconnect task whether or not there is a `slot` and if there is NOT a slot, we finish the disconnect ourselves and fully clean up the connection as if the disconnection fully happened. If there IS a slot, then we just cancel like normal because the disconnect task will be properly scheduled and we can just let it clean itself up like normal.
- If we do not perform the disconnect ourselves and fully clean up the connection, then we get a DEADLOCK forever as the client does not A) have it's connection ref count reduced to 0 properly and B) it does not every invoke the disconnection customer callback, so any code the customer may have will wait forever for a disconnect that will never arrive.
- Note that I tried finishing the disconnect from the disconnect itself, but due to how the tasks are laid out, doing this always leads to segfaults.
- Annoyingly, we cannot "properly" disconnect from the disconnect function itself because we do not have a slot to schedule a task on and, even if we did, we'd be racing with the already scheduled reconnect task...
- If there IS a `slot` and the disconnect is called while reconnecting, then just canceling the reconnect without calling the disconnect callback will work fine, because the disconnect task will be properly scheduled on the `slot` in `aws_mqtt_client_connection_disconnect`.
- If the reconnect task happens BEFORE the disconnect that is also fine, it will just attempt to reconnect because the state will not be `AWS_MQTT_CLIENT_STATE_DISCONNECTING`.

This change should do everything we want and I'm decently confident now that it is the right solution. It should also not change the existing functionality for customers at all and should act as they expect.